core/lib/authorization/capabilities.ts

```diff
@@ -0,0 +1,211 @@
+import type { CommunitiesId, FormsId, PubsId, StagesId, UsersId } from "db/public";
+import { Capabilities } from "db/src/public/Capabilities";
+import { MembershipType } from "db/src/public/MembershipType";
+
+import { db } from "~/kysely/database";
+
+export const pubCapabilities = [
+	Capabilities.movePub,
+	Capabilities.createPub,
+	Capabilities.viewPub,
+	Capabilities.deletePub,
+	Capabilities.updatePubValues,
+	Capabilities.createRelatedPub,
+	Capabilities.editPubWithForm,
+	Capabilities.addPubMember,
+	Capabilities.removePubMember,
+	Capabilities.runAction,
+] as const;
+
+export const communityCapabilities = [
+	Capabilities.createPubField,
+	Capabilities.archivePubField,
+	Capabilities.editPubField,
+	Capabilities.createPubType,
+	Capabilities.editPubType,
+	Capabilities.deletePubType,
+	Capabilities.createStage,
+	Capabilities.addCommunityMember,
+	Capabilities.removeCommunityMember,
+	Capabilities.manageMemberGroups,
+	Capabilities.editCommunity,
+	Capabilities.createForm,
+	Capabilities.createApiToken,
+	Capabilities.revokeApiToken,
+] as const;
+export const stageCapabilities = [
+	Capabilities.viewStage,
+	Capabilities.manageStage,
+	Capabilities.deleteStage,
+	Capabilities.addStageMember,
+	Capabilities.removeStageMember,
+] as const;
+export const formCapabilities = [
+	Capabilities.createPubWithForm,
+	Capabilities.editPubWithForm,
+	Capabilities.addFormMember,
+	Capabilities.removeFormMember,
+	Capabilities.editForm,
+	Capabilities.archiveForm,
+] as const;
+
+type CapabilitiesArg = {
+	[MembershipType.pub]: typeof pubCapabilities;
+	[MembershipType.stage]: typeof stageCapabilities;
+	[MembershipType.community]: typeof communityCapabilities;
+	[MembershipType.form]: typeof formCapabilities;
+};
+
+type Target = PubTarget | StageTarget;
+
+type PubTarget = {
+	type: MembershipType.pub;
+	pubId: PubsId;
+};
+
+type StageTarget = {
+	type: MembershipType.stage;
+	stageId: StagesId;
+};
+
+type CommunityTarget = {
+	type: MembershipType.community;
+	communityId: CommunitiesId;
+};
+
+type FormTarget = {
+	type: MembershipType.form;
+	formId: FormsId;
+};
+
+export const userCan = async <T extends Target>(
+	capability: CapabilitiesArg[T["type"]][number],
+	target: T,
+	userId: UsersId
+) => {
+	if (target.type === MembershipType.pub) {
+		const capabilitiesQuery = db
+			.with("stage", (db) =>
+				db
+					.selectFrom("PubsInStages")
+					.where("PubsInStages.pubId", "=", target.pubId)
+					.select("PubsInStages.stageId")
+			)
+			.with("community", (db) =>
+				db.selectFrom("pubs").where("pubs.id", "=", target.pubId).select("pubs.communityId")
+			)
```

Comment:
For a pub, we first get its stage and community because pub capabilities can be granted by stage or community memberships. Stage capabilities can only be granted by stage or community membership so that query below is slightly simpler (like the community and forms versions will be).